// @volar/vue-language-service/out/documentFeatures/format.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.register = void 0;
const vscode_languageserver_textdocument_1 = require("vscode-languageserver-textdocument");
const vscode = require("vscode-languageserver-protocol");
function register(context) {
    return (document, options) => __awaiter(this, void 0, void 0, function* () {
        var _a;
        const originalDocument = document;
        const rootEdits = yield tryFormat(document);
        const vueDocument = context.getVueDocument(document);
        if (!vueDocument)
            return rootEdits;
        if (rootEdits === null || rootEdits === void 0 ? void 0 : rootEdits.length) {
            applyEdits(rootEdits);
        }
        let level = 0;
        while (true) {
            tryUpdateVueDocument();
            const embeddeds = getEmbeddedsByLevel(vueDocument, level++);
            if (embeddeds.length === 0)
                break;
            let edits = [];
            let toPatchIndent;
            for (const embedded of embeddeds) {
                if (!((_a = embedded.self) === null || _a === void 0 ? void 0 : _a.file.capabilities.formatting))
                    continue;
                const sourceMap = vueDocument.sourceMapsMap.get(embedded.self);
                if (embedded.inheritParentIndent)
                    toPatchIndent = {
                        sourceMapEmbeddedDocumentUri: sourceMap.mappedDocument.uri,
                    };
                const _edits = yield tryFormat(sourceMap.mappedDocument);
                if (!_edits)
                    continue;
                for (const textEdit of _edits) {
                    for (const [range] of sourceMap.getSourceRanges(textEdit.range.start, textEdit.range.end)) {
                        edits.push({
                            newText: textEdit.newText,
                            range,
                        });
                    }
                }
            }
            if (edits.length > 0) {
                applyEdits(edits);
            }
            if (toPatchIndent) {
                tryUpdateVueDocument();
                const sourceMap = vueDocument.getSourceMaps().find(sourceMap => sourceMap.mappedDocument.uri === (toPatchIndent === null || toPatchIndent === void 0 ? void 0 : toPatchIndent.sourceMapEmbeddedDocumentUri));
                if (sourceMap) {
                    const indentEdits = patchInterpolationIndent(vueDocument, sourceMap);
                    if (indentEdits.length > 0) {
                        applyEdits(indentEdits);
                    }
                }
            }
        }
        if (document.getText() === originalDocument.getText())
            return;
        const editRange = vscode.Range.create(originalDocument.positionAt(0), originalDocument.positionAt(originalDocument.getText().length));
        const textEdit = vscode.TextEdit.replace(editRange, document.getText());
        return [textEdit];
        function tryUpdateVueDocument() {
            if ((vueDocument === null || vueDocument === void 0 ? void 0 : vueDocument.getDocument().getText()) !== document.getText()) {
                vueDocument === null || vueDocument === void 0 ? void 0 : vueDocument.file.update(document.getText(), document.version.toString());
            }
        }
        function getEmbeddedsByLevel(vueDocument, level) {
            const embeddeds = vueDocument.file.getEmbeddeds();
            const embeddedsLevels = [embeddeds];
            while (true) {
                if (embeddedsLevels.length > level)
                    return embeddedsLevels[level];
                let nextEmbeddeds = [];
                for (const embeddeds of embeddedsLevels[embeddedsLevels.length - 1]) {
                    nextEmbeddeds = nextEmbeddeds.concat(embeddeds.embeddeds);
                }
                embeddedsLevels.push(nextEmbeddeds);
            }
        }
        function tryFormat(document) {
            return __awaiter(this, void 0, void 0, function* () {
                const plugins = context.getFormatPlugins();
                const range = {
                    start: document.positionAt(0),
                    end: document.positionAt(document.getText().length),
                };
                context.updateTsLs(document);
                for (const plugin of plugins) {
                    if (!plugin.format)
                        continue;
                    let edits;
                    try {
                        edits = yield plugin.format(document, range, options);
                    }
                    catch (err) {
                        console.error(err);
                    }
                    if (!edits)
                        continue;
                    return edits;
                }
            });
        }
        function applyEdits(textEdits) {
            const newText = vscode_languageserver_textdocument_1.TextDocument.applyEdits(document, textEdits);
            if (newText !== document.getText()) {
                document = vscode_languageserver_textdocument_1.TextDocument.create(document.uri, document.languageId, document.version + 1, newText);
            }
        }
    });
}
exports.register = register;
function patchInterpolationIndent(vueDocument, sourceMap) {
    const indentTextEdits = [];
    const document = vueDocument.getDocument();
    for (const mapped of sourceMap.mappings) {
        const textRange = {
            start: document.positionAt(mapped.sourceRange.start),
            end: document.positionAt(mapped.sourceRange.end),
        };
        const text = document.getText(textRange);
        if (text.indexOf('\n') === -1)
            continue;
        const lines = text.split('\n');
        const removeIndent = getRemoveIndent(lines);
        const baseIndent = getBaseIndent(mapped.sourceRange.start);
        for (let i = 1; i < lines.length; i++) {
            const line = lines[i];
            if (line.startsWith(removeIndent)) {
                lines[i] = line.replace(removeIndent, baseIndent);
            }
            else {
                lines[i] = baseIndent.replace(removeIndent, '') + line;
            }
        }
        indentTextEdits.push({
            newText: lines.join('\n'),
            range: textRange,
        });
    }
    return indentTextEdits;
    function getRemoveIndent(lines) {
        const lastLine = lines[lines.length - 1];
        return lastLine.substring(0, lastLine.length - lastLine.trimStart().length);
    }
    function getBaseIndent(pos) {
        const startPos = document.positionAt(pos);
        const startLineText = document.getText({ start: startPos, end: { line: startPos.line, character: 0 } });
        return startLineText.substring(0, startLineText.length - startLineText.trimStart().length);
    }
}
